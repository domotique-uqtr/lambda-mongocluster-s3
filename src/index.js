'use strict';

const AWS = require('aws-sdk');
const fs = require('fs');
const archiver = require('archiver');
const dayjs = require('dayjs');
const exec = require('child_process').exec;

// ENVIRONMENT VARIABLES
const dumpOptions = process.env.MONGODUMP_OPTIONS;
const bucketName = process.env.S3_BUCKET;
const s3bucket = new AWS.S3({ params: { Bucket: bucketName } });
const s3StorageClass = process.env.S3_STORAGE_CLASS || 'STANDARD';
const zipFilename = process.env.ZIP_FILENAME || 'mongodb_backup';
const dateFormat = process.env.DATE_FORMAT || 'YYYYMMDD_HHmmss';

module.exports.handler = function (_event, _context, _cb) {
  console.log(`MongoDB backup to S3 bucket '${bucketName}' is starting`);
  process.env.PATH = process.env.PATH + ':' + process.env.LAMBDA_TASK_ROOT;
  const fileName = zipFilename + '_' + dayjs().format(dateFormat);
  const folderName = `/tmp/${fileName}/`;
  const filePath = `/tmp/${fileName}.zip`;

  exec(
    `mongodump ${dumpOptions} --out ${folderName}`,
    (error, _stdout, _stderr) => {
      if (error) {
        console.log('Mongodump failed: ', error);
        return;
      }

      const output = fs.createWriteStream(filePath);
      const zipArchive = archiver('zip');

      zipArchive.on('warning', function (err) {
        console.log('ZIP warning: ', err);
      });

      zipArchive.on('error', function (err) {
        console.log('ZIP error: ', err);
      });

      output.on('close', function () {
        fs.readFile(filePath, function (err, data) {
          if (err) {
            console.log('readFile failed: ', err);
            return;
          }
          s3bucket.upload(
            {
              Key: fileName + '.zip',
              Body: data,
              ContentType: 'application/zip',
              ServerSideEncryption: 'AES256',
              StorageClass: s3StorageClass,
            },
            function (err, _data) {
              fs.unlink(filePath, function (err) {
                if (err) {
                  console.log('Could not delete temp file: ', err);
                }
              });
              if (err) {
                console.log('Upload to S3 failed: ', err);
              } else {
                console.log('Backup completed successfully');
              }
            },
          );
        });
      });

      zipArchive.pipe(output);
      zipArchive.directory(folderName, false);
      zipArchive.finalize();
    },
  );
};
