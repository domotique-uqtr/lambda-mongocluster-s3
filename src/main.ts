import * as path from 'path';
import * as aws_events from '@aws-cdk/aws-events';
import * as aws_events_targets from '@aws-cdk/aws-events-targets';
import * as lambda from '@aws-cdk/aws-lambda-nodejs';
import * as s3 from '@aws-cdk/aws-s3';
import * as cdk from '@aws-cdk/core';

const ENV = 'habilhome';
const TAGS: { [key: string]: string } = {
  Maintainer: 'Guillaume Maka',
  MaintainerEmail: 'guillaume.maka@gmail.com',
  Project: 'ProjetOntario',
  CostCenter: 'Lachapelle',
};

function _addTags(scope: cdk.Construct) {
  for (const tag in TAGS) {
    cdk.Tags.of(scope).add(tag, TAGS[tag]);
  }
}

function _id(id: string): string {
  return `${ENV}/${id}`;
}

function _exportName(exportName: string): string {
  return `${ENV}::${exportName}`;
}
export class MongoDBBackupLambda extends cdk.Stack {
  constructor(scope: cdk.Construct, id: string, props: cdk.StackProps = {}) {
    super(scope, id, props);

    const fn = new lambda.NodejsFunction(this, _id('MongoDBBackupLambda'), {
      bundling: {
        commandHooks: {
          beforeInstall: (inputDir: string, outputDir: string) => [
            `echo [DEBUG][beforeInstall] inputDir: ${inputDir}, outputDir: ${outputDir}`,
          ],
          afterBundling: (inputDir: string, outputDir: string) => [
            `echo [DEBUG][afterBundling] inputDir: ${inputDir}, outputDir: ${outputDir}`,
          ],
          beforeBundling: (inputDir: string, outputDir: string) => {
            return [
              `echo [DEBUG][beforeBundling] inputDir: ${inputDir}, outputDir: ${outputDir}`,
              `cp ${inputDir}/src/mongodump ${outputDir}`,
            ];
          },
        },
        minify: true,
      },
      entry: path.resolve(__dirname, 'index.js'),
      timeout: cdk.Duration.seconds(120),
      environment: {
        NODE_ENV: 'production',
        MONGODUMP_OPTIONS: `--uri ${cdk.SecretValue.secretsManager(
          'habilhome/prod/MongoDBAtlas',
        )}`,
        S3_BUCKET: cdk.Fn.importValue(_exportName('BucketName')),
        ZIP_FILENAME: 'habilhome-mongodb-backup',
      },
    });

    const s3Bucket = s3.Bucket.fromBucketName(
      this,
      _id('BucketName'),
      cdk.Fn.importValue(_exportName('BucketName')),
    );

    s3Bucket.grantWrite(fn);

    const schedule = aws_events.Schedule.rate(cdk.Duration.days(90));

    new aws_events.Rule(this, _id('MongoDBBackupLambdaRule'), {
      schedule,
      targets: [new aws_events_targets.LambdaFunction(fn)],
    });

    _addTags(this);
  }
}

// for development, use account/region from cdk cli
const env = {
  account:
    process.env.AWS_ACCOUNT_ID ||
    process.env.CDK_DEFAULT_ACOUNT ||
    '650988679566',
  region:
    process.env.AWS_DEFAULT_REGION ||
    // process.env.CDK_DEFAULT_ACOUNT ||
    'ca-central-1', //process.env.AWS_REGION
};

const app = new cdk.App();

new MongoDBBackupLambda(app, 'mongodb-backup-lambda', { env: env });
// new MyStack(app, 'my-stack-prod', { env: prodEnv });

app.synth();
