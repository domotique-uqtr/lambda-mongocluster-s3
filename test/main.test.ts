import '@aws-cdk/assert/jest';
import { App } from '@aws-cdk/core';
import { MongoDBBackupLambda } from '../src/main';

test('Snapshot', () => {
  const app = new App();
  const stack = new MongoDBBackupLambda(app, 'test');

  expect(stack).toHaveResource('AWS::Lambda::Function');
  expect(stack).toHaveResource('AWS::Lambda::Permission');
  expect(stack).toHaveResource('AWS::IAM::Role');
  expect(stack).toHaveResource('AWS::IAM::Policy');
  expect(stack).toHaveResource('AWS::Events::Rule');
  expect(
    app.synth().getStackArtifact(stack.artifactId).template,
  ).toMatchSnapshot();
});
