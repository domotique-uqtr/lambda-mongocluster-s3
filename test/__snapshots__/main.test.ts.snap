// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
Object {
  "Parameters": Object {
    "AssetParametersac048517657d0df80a163e651f888d736c469ef4b36f506b689aa3acaba1d5c7ArtifactHash148AF815": Object {
      "Description": "Artifact hash for asset \\"ac048517657d0df80a163e651f888d736c469ef4b36f506b689aa3acaba1d5c7\\"",
      "Type": "String",
    },
    "AssetParametersac048517657d0df80a163e651f888d736c469ef4b36f506b689aa3acaba1d5c7S3Bucket221F8D3B": Object {
      "Description": "S3 bucket for asset \\"ac048517657d0df80a163e651f888d736c469ef4b36f506b689aa3acaba1d5c7\\"",
      "Type": "String",
    },
    "AssetParametersac048517657d0df80a163e651f888d736c469ef4b36f506b689aa3acaba1d5c7S3VersionKeyA94632E0": Object {
      "Description": "S3 key for asset version \\"ac048517657d0df80a163e651f888d736c469ef4b36f506b689aa3acaba1d5c7\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "habilhomeMongoDBBackupLambda9F6BB278": Object {
      "DependsOn": Array [
        "habilhomeMongoDBBackupLambdaServiceRoleDefaultPolicy98FDE1FE",
        "habilhomeMongoDBBackupLambdaServiceRoleE0740638",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParametersac048517657d0df80a163e651f888d736c469ef4b36f506b689aa3acaba1d5c7S3Bucket221F8D3B",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersac048517657d0df80a163e651f888d736c469ef4b36f506b689aa3acaba1d5c7S3VersionKeyA94632E0",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParametersac048517657d0df80a163e651f888d736c469ef4b36f506b689aa3acaba1d5c7S3VersionKeyA94632E0",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "MONGODUMP_OPTIONS": "--uri {{resolve:secretsmanager:habilhome/prod/MongoDBAtlas:SecretString:::}}",
            "NODE_ENV": "production",
            "S3_BUCKET": Object {
              "Fn::ImportValue": "habilhome::BucketName",
            },
            "ZIP_FILENAME": "habilhome-mongodb-backup",
          },
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "habilhomeMongoDBBackupLambdaServiceRoleE0740638",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Tags": Array [
          Object {
            "Key": "CostCenter",
            "Value": "Lachapelle",
          },
          Object {
            "Key": "Maintainer",
            "Value": "Guillaume Maka",
          },
          Object {
            "Key": "MaintainerEmail",
            "Value": "guillaume.maka@gmail.com",
          },
          Object {
            "Key": "Project",
            "Value": "ProjetOntario",
          },
        ],
        "Timeout": 120,
      },
      "Type": "AWS::Lambda::Function",
    },
    "habilhomeMongoDBBackupLambdaRule7809CB1D": Object {
      "Properties": Object {
        "ScheduleExpression": "rate(90 days)",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "habilhomeMongoDBBackupLambda9F6BB278",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "habilhomeMongoDBBackupLambdaRuleAllowEventRuletesthabilhomeMongoDBBackupLambda59DFFA960BBD3A80": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "habilhomeMongoDBBackupLambda9F6BB278",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "habilhomeMongoDBBackupLambdaRule7809CB1D",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "habilhomeMongoDBBackupLambdaServiceRoleDefaultPolicy98FDE1FE": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:DeleteObject*",
                "s3:PutObject*",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Fn::ImportValue": "habilhome::BucketName",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Fn::ImportValue": "habilhome::BucketName",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "habilhomeMongoDBBackupLambdaServiceRoleDefaultPolicy98FDE1FE",
        "Roles": Array [
          Object {
            "Ref": "habilhomeMongoDBBackupLambdaServiceRoleE0740638",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "habilhomeMongoDBBackupLambdaServiceRoleE0740638": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
        "Tags": Array [
          Object {
            "Key": "CostCenter",
            "Value": "Lachapelle",
          },
          Object {
            "Key": "Maintainer",
            "Value": "Guillaume Maka",
          },
          Object {
            "Key": "MaintainerEmail",
            "Value": "guillaume.maka@gmail.com",
          },
          Object {
            "Key": "Project",
            "Value": "ProjetOntario",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
